sql/analyzer/select_hints.go

```diff
@@ -0,0 +1,452 @@
+	HintTypeLookupJoin                               // LOOKUP_JOIN
+	HintTypeHashJoin                                 // HASH_JOIN
+	HintTypeSemiJoin                                 // SEMI_JOIN
+	HintTypeAntiJoin                                 // ANTI_JOIN
+	HintTypeInnerJoin                                // INNER_JOIN
+	HintTypeNoIndexConditionPushDown                 // NO_ICP
+	HintTypeMaxExecutionTime                         // MAX_EXECUTION_TIME
+	HintTypeSetVar                                   // SET_VAR
+)
+
+type Hint struct {
+	Typ  HintType
+	Args []string
+}
+
+func (h Hint) String() string {
+	if len(h.Args) > 0 {
+		return fmt.Sprintf("%s(%s)", h.Typ, strings.Join(h.Args, ","))
+	} else {
+		return h.Typ.String()
+	}
+}
+
+func newHint(joinTyp string, args []string) Hint {
+	var typ HintType
+	switch joinTyp {
+	case "join_order":
+		typ = HintTypeJoinOrder
+	case "join_fixed_order":
+		typ = HintTypeJoinFixedOrder
+	case "merge_join":
+		typ = HintTypeMergeJoin
+	case "lookup_join":
+		typ = HintTypeLookupJoin
+	case "hash_join":
+		typ = HintTypeHashJoin
+	case "inner_join":
+		typ = HintTypeInnerJoin
+	case "semi_join":
+		typ = HintTypeSemiJoin
+	case "anti_join":
+		typ = HintTypeAntiJoin
+	case "no_icp":
+		typ = HintTypeNoIndexConditionPushDown
+	case "max_execution_time":
+		typ = HintTypeMaxExecutionTime
+	case "set_var":
+		typ = HintTypeSetVar
+	default:
+		typ = HintTypeUnknown
+	}
+	return Hint{Typ: typ, Args: args}
+}
+
+func (h Hint) valid() bool {
+	switch h.Typ {
+	case HintTypeJoinOrder:
+		return len(h.Args) > 0
+	case HintTypeJoinFixedOrder:
+		return len(h.Args) == 0
+	case HintTypeMergeJoin:
+		return len(h.Args) == 2
+	case HintTypeLookupJoin:
+		return len(h.Args) == 2
+	case HintTypeHashJoin:
+		return len(h.Args) == 2
+	case HintTypeInnerJoin:
+		return len(h.Args) == 2
+	case HintTypeSemiJoin:
+		return len(h.Args) == 2
+	case HintTypeAntiJoin:
+		return len(h.Args) == 2
+	case HintTypeNoIndexConditionPushDown:
+		return len(h.Args) == 0
+	case HintTypeMaxExecutionTime:
+		return len(h.Args) == 0
+	case HintTypeSetVar:
+		return len(h.Args) == 0
+	case HintTypeUnknown:
+	default:
+	}
+	return true
+}
+
+var hintRegex = regexp.MustCompile("(\\s*[a-z_]+(\\([^\\(]+\\))?\\s*)")
+
+type hintState uint8
+
+const (
+	hsUnknown hintState = iota
+	hsNextHint
+	hsGetType
+	hsGetArgs
+	hsFinalizeHint
+)
+
+func extractJoinHint(n *plan.JoinNode) []Hint {
+	if n.Comment() != "" {
+		return parseJoinHints(n.Comment())
+	}
+	return nil
+}
+
+// TODO: this is pretty nasty. Should be done in the parser instead.
+func parseJoinHints(comment string) []Hint {
+	comment = strings.TrimPrefix(comment, "/*+")
+	comment = strings.TrimSuffix(comment, "*/")
+	comment = strings.ToLower(strings.TrimSpace(comment))
+	comments := hintRegex.FindAll([]byte(comment), -1)
+	var hints []Hint
+	state := hsNextHint
+	var hintStr string
+	var hintType string
+	var args []string
+	for {
```

Comment:
Typically you would use regex capture groups to do this. If you're going to bother using regexes here, you should probably use them this way.

Rather than regex.FindAll, use regex.Find to get hints one at a time, then pull the name of the hint and the tables from the capture groups. Going to be a lot less fragile (and easier to read) than this.